File: scripts/mk_parts_recipes.rb
#!env ruby

# frozen_string_literal: true

require "nokogiri"

config_dir = "/mnt/s/Games/Steam/steamapps/common/7 Days To Die/Data/Config"
items_file = "#{config_dir}/items.xml"

raise "#{config_dir} Does not exist" unless File.exist?(config_dir) && File.directory?(config_dir)

items = Nokogiri::XML(File.read(items_file))

output = Nokogiri::XML::Builder.new do |xml|
  xml.configs do
    xml.append(xpath: "/recipes") do
      items.xpath("//item").each do |item|
        recipe_name = item.at_xpath("@name").to_s

        next unless recipe_name.match?(/(?:armor|gun|melee).*Parts$/)

        metal = recipe_name.match?(/T[1-2].*Parts$/) ? "resourceForgedIron" : "resourceForgedSteel"
        parts = recipe_name.match?(/(Robotics|StunBaton)Parts$/) ? "resourceElectricParts" : "resourceMechanicalParts"

        xml.recipe(name: recipe_name, count: 1, craft_area: "workbench", tags: "workbenchCrafting") do
          xml.ingredient(name: metal, count: 1)
          xml.ingredient(name: "resourceDuctTape", count: 5)
          xml.ingredient(name: parts, count: 10)
        end
      end
    end
  end
end

puts output.to_xml indent: 2, save_with: Nokogiri::XML::Node::SaveOptions::DEFAULT_XML | Nokogiri::XML::Node::SaveOptions::NO_DECLARATION

File: scripts/mk_parts_recipes.py
#!/usr/bin/env python3
"""Regenerate the craftableparts recipes from the game's items.xml.

Writes the recipes XML to stdout; mkall.sh redirects it into
modlets/donovan-craftableparts/Config/recipes.xml.
"""

import re
import sys
from pathlib import Path

from lxml import etree

# The repo's Config symlink points at the game's Data/Config directory
config_dir = Path(__file__).resolve().parent.parent / "Config"
items_file = config_dir / "items.xml"

if not config_dir.is_dir():
    sys.exit(f"{config_dir} Does not exist")

items = etree.parse(str(items_file))

configs = etree.Element("configs")
append = etree.SubElement(configs, "append", xpath="/recipes")

for item in items.xpath("//item"):
    recipe_name = item.get("name", "")

    if not re.search(r"(?:armor|gun|melee).*Parts$", recipe_name):
        continue

    metal = ("resourceForgedIron" if re.search(r"T[1-2].*Parts$", recipe_name)
             else "resourceForgedSteel")
    parts = ("resourceElectricParts" if re.search(r"(Robotics|StunBaton)Parts$", recipe_name)
             else "resourceMechanicalParts")

    recipe = etree.SubElement(
        append, "recipe",
        name=recipe_name, count="1",
        craft_area="workbench", tags="workbenchCrafting")

    etree.SubElement(recipe, "ingredient", name=metal, count="1")
    etree.SubElement(recipe, "ingredient", name="resourceDuctTape", count="5")
    etree.SubElement(recipe, "ingredient", name=parts, count="10")

etree.indent(configs, space="  ")
print(etree.tostring(configs).decode())
